pkg/controller/endpointslice/topologycache/topologycache.go

```diff
@@ -129,19 +142,36 @@ func (t *TopologyCache) AddHints(si *SliceInfo) ([]*discovery.EndpointSlice, []*
 		allocatedHintsByZone[zone] += diff
 	}
 
+	if len(allocatedHintsByZone) == 0 {
+		klog.V(2).InfoS("No hints allocated for zones, removing them", "serviceKey", si.ServiceKey, "addressType", si.AddressType)
+		events = append(events, &EventBuilder{
+			EventType: v1.EventTypeWarning,
+			Reason:    "TopologyAwareHintsDisabled",
+			Message:   FormatWithAddressType(NoAllocatedHintsForZones, si.AddressType),
+		})
+		t.RemoveHints(si.ServiceKey, si.AddressType)
+		slicesToCreate, slicesToUpdate := RemoveHintsFromSlices(si)
+		return slicesToCreate, slicesToUpdate, events
+	}
+
+	hintsEnabled := t.hintsPopulatedByService.Has(si.ServiceKey)
 	t.SetHints(si.ServiceKey, si.AddressType, allocatedHintsByZone)
-	return si.ToCreate, si.ToUpdate
+
+	// if hints were not enabled before, we publish an event to indicate we enabled them.
+	if !hintsEnabled {
+		klog.InfoS("Topology Aware Hints has been enabled, adding hints.", "serviceKey", si.ServiceKey, "addressType", si.AddressType)
+		events = append(events, &EventBuilder{
+			EventType: v1.EventTypeNormal,
+			Reason:    "TopologyAwareHintsEnabled",
+			Message:   FormatWithAddressType(TopologyAwareHintsEnabled, si.AddressType),
+		})
+	}
+	return si.ToCreate, si.ToUpdate, events
 }
 
 // SetHints sets topology hints for the provided serviceKey and addrType in this
 // cache.
 func (t *TopologyCache) SetHints(serviceKey string, addrType discovery.AddressType, allocatedHintsByZone EndpointZoneInfo) {
-	if len(allocatedHintsByZone) == 0 {
```

Comment:
is this safe now?

Reply:
This was moved from SetHints to AddHints. The only place SetHints is called is from AddHints, so it's not a significant difference, but the diff here is definitely confusing.

Reply:
yeah I moved it to AddHints.